solver/src/solver/baseline_solver.rs

```diff
@@ -210,13 +205,28 @@ impl BaselineSolver {
                 let best = candidates
                     .iter()
                     .filter_map(|path| estimate_sell_amount(order.buy_amount, path, amms))
+                    .filter(|estimate| estimate.value <= order.sell_amount)
+                    // For buy orders we find the best path starting at the buy token ending at the
+                    // sell token. When we turn this into a settlement however we need to go from
+                    // the sell token to the buy token. This reversing of the direction can fail or
+                    // yield different amounts as explained in the BaselineSolvable trait.
+                    .filter(|estimate| {
+                        matches!(
+                            traverse_path_forward(
+                                order.sell_token,
+                                estimate.value,
+                                &estimate.path,
+                            ), Some(amount) if amount >= order.buy_amount
```

Comment:
Is this check really necessary? Wouldn't - if the computed buy amount was slightly lower than the order buy amount - the remainder just be filled from our internal balances as long as the prices announced in the solution satisfy the user's limit price?

Also, wouldn't the equivalent of checking `estimate.value >= order.buy_amount` below, be verifying that ` `estimate.value <= order.sell_amount` here (but I assume this limit price check is also done somewhere else)?

Reply:
`traverse_path_forward` succeeding is necessary.  `estimate.value <= order.sell_amount` is the equivalent of what happens for OrderKind::Sell. Here it happens in the line above this `.filter` (but hard to spot in the diff, i managed to miss it at first now too and thought i hadn't committed it).
I still think it makes sense to also have `amount >= order.buy_amount` as we are supposed to produce a correct settlement here which should work even without buffers.